### DISCUSSION TITLE
Implementing / improving list_intersect

### DISCUSSION THREAD

**Discussion by nschagen** [duckdb/duckdb#8178] [Q&A]
Dear duckdb devs,

I've recently been tinkering with the code and was trying to add support for `list_intersect` since I am building an application that heavily relies on it. But I noticed that it was very recently merged but has not appeared in a release yet.  https://github.com/duckdb/duckdb/pull/7518.  When is this expected to be released?

I made a start implementing the `list_intersect` function from scratch instead of using a marco. This is a rather slow process as I normally write python and am still learning the duckdb codebase. My goal was to implement the function using simple nested loops, which would give similar performance (M * N) as the current macro solution.

```
// FIXME implement as actual function if we encounter a lot of performance issues. Complexity now: n * m, with hashing possibly n + m
    {DEFAULT_SCHEMA, "list_intersect", {"l1", "l2", nullptr}, "list_filter(l1, (x) -> list_contains(l2, x))"},
```

This comment suggests that it might be doable to improve the performance once we have a full `ScalarFunction` implementation for it. How can this be done potentially? That would give me a reason to continue my efforts and would also improve `list_has_any` and `list_has_all` performance.

The comment above mentions hashing possibility but I couldn't find out where this hashing actually happens. Any ideas?

Thanks in advance!

**Comment by Tishj** [comment]
@maiadegraaf maybe you could give some quick pointers if you have time?

**Comment by taniabogatsch** [comment]
Hi, @nschagen! We try to have a release roughly every three months, and the rewrites will be in the next release.

As for better performance, you can hash one of the two lists (l1) in n by hashing each element. Then, you can go through the other list (l2) in m and use the hash function to probe against the first list. This is very similar to the theory behind hash joins, for example.

**Reply by taniabogatsch** [reply]
You are welcome to work on this! We're always open to PRs. But if you need more support/guidance on this from the developers, and also on implementation details, then it might be interesting for you to reach out to [DuckDB Labs](https://duckdblabs.com/contact/contracting/). Just leaving it here, in case. 😄

**Comment by NickCrews** [comment]
[Relevant code](https://github.com/duckdb/duckdb/blob/c5182bc3b952ebbb90eb721039577ffef4586965/src/catalog/default/default_functions.cpp#L112-L114)
